cabal-install/src/Distribution/Client/ProjectConfig/Legacy.hs

```diff
@@ -245,41 +247,51 @@ parseProject
 parseProject rootPath cacheDir httpTransport verbosity configToParse = do
   let (dir, projectFileName) = splitFileName rootPath
   projectDir <- makeAbsolute dir
-  projectPath <- canonicalizeConfigPath projectDir (ProjectConfigPath $ projectFileName :| [])
-  parseProjectSkeleton cacheDir httpTransport verbosity projectDir projectPath configToParse
+  projectPath@(ProjectConfigPath (canonicalRoot :| _)) <- canonicalizeConfigPath projectDir (ProjectConfigPath $ projectFileName :| [])
+  importsBy <- newIORef [(canonicalRoot, projectPath)]
+  parseProjectSkeleton cacheDir httpTransport verbosity importsBy projectDir projectPath configToParse
 
 parseProjectSkeleton
   :: FilePath
   -> HttpTransport
   -> Verbosity
+  -> IORef [(FilePath, ProjectConfigPath)]
+  -- ^ The imports seen so far, used to report on cycles and duplicates and to detect duplicates that are not cycles
   -> FilePath
   -- ^ The directory of the project configuration, typically the directory of cabal.project
   -> ProjectConfigPath
   -- ^ The path of the file being parsed, either the root or an import
   -> ProjectConfigToParse
   -- ^ The contents of the file to parse
   -> IO (ParseResult ProjectConfigSkeleton)
-parseProjectSkeleton cacheDir httpTransport verbosity projectDir source (ProjectConfigToParse bs) =
+parseProjectSkeleton cacheDir httpTransport verbosity importsBy projectDir source (ProjectConfigToParse bs) =
   (sanityWalkPCS False =<<) <$> liftPR (go []) (ParseUtils.readFields bs)
   where
     go :: [ParseUtils.Field] -> [ParseUtils.Field] -> IO (ParseResult ProjectConfigSkeleton)
     go acc (x : xs) = case x of
       (ParseUtils.F _ "import" importLoc) -> do
         let importLocPath = importLoc `consProjectConfigPath` source
 
-        -- Once we canonicalize the import path, we can check for cyclical imports
-        normLocPath <- canonicalizeConfigPath projectDir importLocPath
+        -- Once we canonicalize the import path, we can check for cyclical and duplicate imports
+        normLocPath@(ProjectConfigPath (uniqueImport :| _)) <- canonicalizeConfigPath projectDir importLocPath
+        seenImportsBy@(fmap fst -> seenImports) <- atomicModifyIORef' importsBy (\ibs -> (nub $ (uniqueImport, normLocPath) : ibs, ibs))
```

Comment:
Drive-by-comment, but isn't this `(nub $ (uniqueImport, normLocPath) : ibs` in O(n³)?

I suppose, usually there are at most 100 packages that import each other, wouldn't a `Set` be better here, since it looks like the order doesn't really matter any way?

Reply:
Already better is `ordNub` but I see we have `NubList`. Having a look at that now.

https://github.com/haskell/cabal/blob/03d98294e1c071be368f04dbe6fc9cb263bea86d/Cabal/src/Distribution/Utils/NubList.hs#L40-L55

Reply:
I'm now using `NubList`.